File: android/src/main/java/expo/modules/libvlcplayer/player/MediaPlayer.kt
package expo.modules.libvlcplayer.player

import android.content.res.AssetFileDescriptor
import android.graphics.Matrix
import android.net.Uri
import android.view.TextureView
import android.view.ViewGroup
import expo.modules.libvlcplayer.LibVlcPlayerView
import expo.modules.libvlcplayer.constants.MediaPlayerConstants
import expo.modules.libvlcplayer.enums.VideoContentFit
import expo.modules.libvlcplayer.managers.MediaPlayerManager
import expo.modules.libvlcplayer.records.Dialog
import expo.modules.libvlcplayer.records.Media
import expo.modules.libvlcplayer.records.MediaInfo
import expo.modules.libvlcplayer.records.MediaTrack
import expo.modules.libvlcplayer.records.MediaTracks
import expo.modules.libvlcplayer.records.Metadata
import expo.modules.libvlcplayer.records.Recording
import expo.modules.libvlcplayer.records.Slave
import expo.modules.libvlcplayer.records.Video
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.videolan.libvlc.LibVLC
import org.videolan.libvlc.MediaPlayer.Event
import org.videolan.libvlc.MediaPlayer.EventListener
import org.videolan.libvlc.interfaces.IMedia
import org.videolan.libvlc.util.DisplayManager
import org.videolan.libvlc.util.VLCVideoLayout
import java.io.File
import java.io.FileOutputStream
import java.net.URI
import org.videolan.libvlc.Dialog as VLCDialog
import org.videolan.libvlc.Media as VLCMedia
import org.videolan.libvlc.MediaPlayer as VLCMediaPlayer

private val DISPLAY_MANAGER: DisplayManager? = null
private val ENABLE_SUBTITLES: Boolean = true
private val USE_TEXTURE_VIEW: Boolean = true

class MediaPlayer(
  val view: LibVlcPlayerView,
) {
  private val context = view.context

  val video: VLCVideoLayout = VLCVideoLayout(context)
  val picture: VLCVideoLayout = VLCVideoLayout(context)

  private var sourceFd: AssetFileDescriptor? = null
  private var pauseJob: Job? = null

  var libVlc: LibVLC? = null
  var mediaPlayer: VLCMediaPlayer? = null
  var vlcDialog: VLCDialog? = null

  var userStop: Boolean = false
  var firstPlay: Boolean = true
  var shouldInit: Boolean = true

  val hasVideoSize: Boolean
    get() {
      val video = getVideo()
      return video.width > 0 && video.height > 0
    }

  val hasMediaLength: Boolean
    get() {
      val length = getLength()
      return length > 0
    }

  val hasMediaVolume: Boolean
    get() {
      val volume = mediaPlayer?.getVolume() ?: MediaPlayerConstants.MIN_PLAYER_VOLUME
      return volume > MediaPlayerConstants.MIN_PLAYER_VOLUME
    }

  fun getSourceId(source: String): Int? {
    if (Uri.parse(source).scheme != null) return null

    val identifier = context.resources.getIdentifier(source, "raw", context.packageName)

    return identifier.takeIf { it != 0 }
  }

  fun openSourceFd(source: String): AssetFileDescriptor? {
    val sourceId = getSourceId(source) ?: return null

    sourceFd =
      try {
        context.resources.openRawResourceFd(sourceId)
      } catch (_: Exception) {
        null
      }

    return sourceFd
  }

  fun createMedia(
    libVlc: LibVLC,
    source: String,
  ): VLCMedia {
    val file = openSourceFd(source)

    return if (file != null) {
      VLCMedia(libVlc, file)
    } else {
      VLCMedia(libVlc, Uri.parse(source))
    }
  }

  fun getSourceUri(source: String): Uri {
    val sourceUri = Uri.parse(source)
    val sourceId = getSourceId(source) ?: return sourceUri
    val cacheDir = File(context.cacheDir, MediaPlayerConstants.SOURCE_CACHE_DIR)
    val file = File(cacheDir, source)

    try {
      cacheDir.mkdirs()
      context.resources.openRawResource(sourceId).use { input ->
        FileOutputStream(file).use { output -> input.copyTo(output) }
      }
    } catch (_: Exception) {
      return sourceUri
    }

    return Uri.fromFile(file)
  }

  fun getTextureView(layout: VLCVideoLayout): TextureView? =
    layout.findViewById(org.videolan.R.id.texture_video)

  fun addPlayerLayout(layout: VLCVideoLayout) {
    val parent = video.parent as? ViewGroup

    if (parent == null) {
      view.addView(layout)
    }
  }

  fun removePlayerView() {
    val parent = video.parent as? ViewGroup

    if (parent != null) {
      view.removeView(video)
    }
  }

  fun resetPlayer() {
    detachPlayer()
    attachPlayer()
  }

  fun attachPlayer() {
    attachPlayerView(video)
    addPlayerLayout(video)
  }

  fun detachPlayer() {
    detachPlayerView()
    removePlayerView()
  }

  fun attachPlayerView(layout: VLCVideoLayout) {
    mediaPlayer?.let { player ->
      val attached = player.getVLCVout().areViewsAttached()

      if (!attached) {
        player.attachViews(layout, DISPLAY_MANAGER, ENABLE_SUBTITLES, USE_TEXTURE_VIEW)
      }
    }
  }

  fun detachPlayerView() {
    mediaPlayer?.let { player ->
      val attached = player.getVLCVout().areViewsAttached()

      if (attached) {
        player.detachViews()
      }
    }
  }

  fun initPlayer() {
    if (shouldInit) {
      destroyPlayer()

      if (view.source != null) {
        createPlayer()
      }
    }
  }

  fun createPlayer() {
    if (view.pictureInPicture) {
      MediaPlayerManager.pictureInPictureManager.setupPipView(view)
    }

    libVlc = LibVLC(context)
    setDialogCallbacks()

    mediaPlayer = VLCMediaPlayer(libVlc!!)
    setPlayerListener()

    attachPlayerView(video)
    addPlayerSlaves(view.slaves)

    try {
      URI(view.source)
    } catch (_: Exception) {
      view.onEncounteredError(mapOf("message" to "Invalid source, media could not be set"))
      return
    }

    var args = view.options
    args.normalizeOptions()
    args.toggleStartPausedOption(view.autoplay)

    val media = createMedia(libVlc!!, view.source!!)
    args.forEach { arg -> media.addOption(arg) }
    mediaPlayer!!.setMedia(media)
    media.release()
    mediaPlayer!!.play()

    firstPlay = true
    shouldInit = false

    addPlayerLayout(video)
  }

  fun destroyPlayer() {
    cancelPauseDelay()
    sourceFd?.close()
    sourceFd = null
    libVlc?.release()
    libVlc = null
    mediaPlayer?.release()
    mediaPlayer = null
    vlcDialog = null
    view.removeAllViews()
  }

  fun setupPlayer() {
    view.post {
      mediaPlayer?.let { player ->
        if (view.scale != MediaPlayerConstants.DEFAULT_PLAYER_SCALE) {
          player.setScale(view.scale.toFloat())
        }

        if (view.rate != MediaPlayerConstants.DEFAULT_PLAYER_RATE) {
          player.setRate(view.rate.toFloat())
        }

        if (view.time != MediaPlayerConstants.DEFAULT_PLAYER_TIME) {
          player.setTime(view.time.toLong())
        }

        // Negative volume workaround
        retryUntil {
          val newVolume =
            if (view.mute) {
              MediaPlayerConstants.MIN_PLAYER_VOLUME
            } else {
              view.volume
            }

          player.setVolume(newVolume)

          return@retryUntil false
        }

        view.time = MediaPlayerConstants.DEFAULT_PLAYER_TIME
      }
    }
  }

  fun addPlayerSlaves(slaves: List<Slave>) {
    slaves.forEach { slave ->
      val source = slave.source
      val type = slave.type
      val slaveType =
        if (type == "subtitle") {
          IMedia.Slave.Type.Subtitle
        } else {
          IMedia.Slave.Type.Audio
        }
      val selected = slave.selected ?: false

      try {
        URI(source)
      } catch (_: Exception) {
        view.onEncounteredError(mapOf("message" to "Invalid source, $type could not be added"))
        return@forEach
      }

      mediaPlayer?.addSlave(slaveType, getSourceUri(source), selected)
    }
  }

  fun selectTrack(
    index: Int,
    type: Int,
  ) {
    mediaPlayer?.let { player ->
      if (index == -1) {
        player.unselectTrackType(type)
      } else {
        player.selectTrack(index.toString())
      }
    }
  }

  fun setPlayerTracks() {
    val audioTrack = view.tracks?.audio
    val videoTrack = view.tracks?.video
    val spuTrack = view.tracks?.subtitle

    audioTrack?.let { track -> selectTrack(track, IMedia.Track.Type.Audio) }
    videoTrack?.let { track -> selectTrack(track, IMedia.Track.Type.Video) }
    spuTrack?.let { track -> selectTrack(track, IMedia.Track.Type.Text) }
  }

  fun setPlayerDelays() {
    mediaPlayer?.let { player ->
      val audioDelay = view.delays?.audio
      val spuDelay = view.delays?.subtitle

      audioDelay?.let { delay -> player.setAudioDelay(delay) }
      spuDelay?.let { delay -> player.setSpuDelay(delay) }
    }
  }

  fun setContentFit(layout: VLCVideoLayout) {
    view.post {
      val textureView = getTextureView(layout) ?: return@post
      val matrix = Matrix()

      val video = getVideo()

      if (hasVideoSize) {
        val viewWidth = textureView.width.toFloat()
        val viewHeight = textureView.height.toFloat()

        val videoWidth = video.width.toFloat()
        val videoHeight = video.height.toFloat()

        val viewAspect = viewWidth / viewHeight
        val videoAspect = videoWidth / videoHeight

        val pivotX = viewWidth / 2f
        val pivotY = viewHeight / 2f

        when (view.contentFit) {
          VideoContentFit.CONTAIN -> {
            // No scaling required
          }

          VideoContentFit.COVER -> {
            val scale =
              if (videoAspect > viewAspect) {
                videoAspect / viewAspect
              } else {
                viewAspect / videoAspect
              }

            matrix.setScale(scale, scale, pivotX, pivotY)
          }

          VideoContentFit.FILL -> {
            var scaleX = 1f
            var scaleY = 1f

            if (videoAspect > viewAspect) {
              scaleY = videoAspect / viewAspect
            } else {
              scaleX = viewAspect / videoAspect
            }

            matrix.setScale(scaleX, scaleY, pivotX, pivotY)
          }
        }
      }

      textureView.setTransform(matrix)
    }
  }

  fun applyContentFit() {
    setContentFit(layout = video)
    setContentFit(layout = picture)
  }

  fun getMediaTracks(): MediaTracks {
    val player = mediaPlayer ?: return MediaTracks()

    val disableTrack = MediaTrack(id = -1, name = "Disable")

    val audios =
      player
        .getTracks(
          IMedia.Track.Type.Audio,
        )?.mapIndexed { index, track -> MediaTrack(id = index, name = track.name) }
    val videos =
      player.getTracks(IMedia.Track.Type.Video)?.mapIndexed {
        index,
        track,
        ->
        MediaTrack(id = index, name = track.name)
      }
    val subtitles =
      player.getTracks(IMedia.Track.Type.Text)?.mapIndexed {
        index,
        track,
        ->
        MediaTrack(id = index, name = track.name)
      }

    val audio = listOf(disableTrack) + (audios ?: emptyList())
    val video = listOf(disableTrack) + (videos ?: emptyList())
    val subtitle = listOf(disableTrack) + (subtitles ?: emptyList())

    return MediaTracks(
      audio = audio,
      video = video,
      subtitle = subtitle,
    )
  }

  fun getMedia(): Media {
    val length = (mediaPlayer?.getLength() ?: 0).toInt()
    val seekable = mediaPlayer?.isSeekable() ?: false

    return Media(
      length = length,
      seekable = seekable,
    )
  }

  fun getMetadata(): Metadata {
    val media = mediaPlayer?.media ?: return Metadata()

    val title = media.getMeta(IMedia.Meta.Title)
    val artist = media.getMeta(IMedia.Meta.Artist)
    val album = media.getMeta(IMedia.Meta.Album)
    val artworkURL = media.getMeta(IMedia.Meta.ArtworkURL)

    return Metadata(
      title = title,
      artist = artist,
      album = album,
      artworkURL = artworkURL,
    )
  }

  fun getVideo(): Video {
    val video =
      mediaPlayer?.getSelectedTrack(IMedia.Track.Type.Video) as? IMedia.VideoTrack ?: return Video()

    val width = video.width
    val height = video.height
    val frameRate =
      if (video.frameRateDen != 0) {
        video.frameRateNum / video.frameRateDen
      } else {
        0
      }
    val bitrate = video.bitrate

    return Video(
      width = width,
      height = height,
      frameRate = frameRate,
      bitrate = bitrate,
    )
  }

  fun getMediaInfo(): MediaInfo {
    val media = getMedia()
    val metadata = getMetadata()
    val video = getVideo()

    return MediaInfo(
      media = media,
      metadata = metadata,
      video = video,
    )
  }

  fun getLength(): Int = (mediaPlayer?.getLength() ?: 0).toInt()

  fun pauseDelay() {
    cancelPauseDelay()

    pauseJob =
      CoroutineScope(Dispatchers.Main).launch {
        delay(MediaPlayerConstants.PAUSE_DELAY_MS)
        mediaPlayer?.pause()
      }
  }

  fun cancelPauseDelay() {
    pauseJob?.cancel()
  }

  fun onStartPictureInPicture() {
    MediaPlayerManager.pictureInPictureManager.layoutForPipEnter()
    view.onPictureInPictureStart(Unit)
  }

  fun onStopPictureInPicture() {
    MediaPlayerManager.pictureInPictureManager.layoutForPipExit()
    view.onPictureInPictureStop(Unit)
  }

  fun retryUntil(
    maxRetries: Int = MediaPlayerConstants.MAX_RETRY_COUNT,
    retry: Int = 0,
    delay: Double = MediaPlayerConstants.RETRY_DELAY_MS,
    block: (isLastAttempt: Boolean) -> Boolean,
  ) {
    val isLastAttempt = retry > maxRetries

    if (block(isLastAttempt) || isLastAttempt) return

    val wait = if (retry > 0) delay else 0.0
    val nextDelay = if (retry > 0) delay * MediaPlayerConstants.EXP_DELAY_MULTIPLIER else delay
    val postDelay = wait.toLong()

    view.postDelayed({
      retryUntil(maxRetries, retry + 1, nextDelay, block)
    }, postDelay)
  }
}

fun MediaPlayer.setPlayerListener() {
  mediaPlayer?.let { player ->
    player.setEventListener(
      EventListener { event ->
        val type = event.type

        when (type) {
          Event.Buffering -> {
            view.onBuffering(mapOf("value" to event.getBuffering()))
          }

          Event.Playing,
          Event.Paused,
          Event.Stopped,
          -> {
            if (type == Event.Playing) {
              view.onPlaying(Unit)

              if (firstPlay) {
                setupPlayer()
                setPlayerTracks()
                setPlayerDelays()

                retryUntil {
                  if (hasVideoSize) {
                    applyContentFit()
                  }

                  return@retryUntil hasVideoSize
                }

                retryUntil {
                  if (hasMediaVolume) {
                    MediaPlayerManager.audioFocusManager.updateAudioFocus()
                  }

                  return@retryUntil hasMediaVolume
                }

                retryUntil { isLastAttempt ->
                  if (hasMediaLength || isLastAttempt) {
                    view.onFirstPlay(getMediaInfo())
                  }

                  return@retryUntil hasMediaLength
                }

                firstPlay = false
              }
            }

            if (type == Event.Paused) {
              view.onPaused(Unit)
            }

            if (type == Event.Stopped) {
              resetPlayer()
              view.onStopped(Unit)

              if (view.repeat && !userStop) {
                player.play()
              }

              userStop = false
              firstPlay = true
            }

            MediaPlayerManager.keepAwakeManager.toggleKeepAwake()
            MediaPlayerManager.audioFocusManager.updateAudioFocus()
            MediaPlayerManager.pictureInPictureManager.setPipActions()
          }

          Event.EndReached -> {
            player.stop()
          }

          Event.EncounteredError -> {
            view.onEncounteredError(mapOf("message" to "Player encountered an error"))
            player.stop()
          }

          Event.TimeChanged -> {
            view.onTimeChanged(mapOf("value" to player.getTime().toInt()))
          }

          Event.PositionChanged -> {
            view.onPositionChanged(mapOf("value" to player.getPosition()))
          }

          Event.ESAdded -> {
            view.onESAdded(getMediaTracks())
          }

          Event.RecordChanged -> {
            val recording =
              Recording(
                path = event.getRecordPath(),
                isRecording = event.getRecording(),
              )

            view.onRecordChanged(recording)
          }
        }
      },
    )
  }
}

fun MediaPlayer.setDialogCallbacks() {
  libVlc?.let { libVlc ->
    VLCDialog.setCallbacks(
      libVlc,
      object : VLCDialog.Callbacks {
        override fun onDisplay(dialog: VLCDialog.ErrorMessage) {
          vlcDialog = dialog

          val dialog =
            Dialog(
              title = dialog.getTitle(),
              text = dialog.getText(),
              type = "error",
            )

          view.onDialogDisplay(dialog)
        }

        override fun onDisplay(dialog: VLCDialog.LoginDialog) {
          vlcDialog = dialog

          val dialog =
            Dialog(
              title = dialog.getTitle(),
              text = dialog.getText(),
              type = "login",
            )

          view.onDialogDisplay(dialog)
        }

        override fun onDisplay(dialog: VLCDialog.QuestionDialog) {
          vlcDialog = dialog

          val dialog =
            Dialog(
              title = dialog.getTitle(),
              text = dialog.getText(),
              type = "question",
              cancelText = dialog.getCancelText(),
              action1Text = dialog.getAction1Text(),
              action2Text = dialog.getAction2Text(),
            )

          view.onDialogDisplay(dialog)
        }

        override fun onDisplay(dialog: VLCDialog.ProgressDialog) {}

        override fun onCanceled(dialog: VLCDialog) {}

        override fun onProgressUpdate(dialog: VLCDialog.ProgressDialog) {}
      },
    )
  }
}

private fun MutableList<String>.normalizeOptions() {
  val normalized =
    map { option ->
      if (!option.startsWith(":")) {
        ":" + option.dropWhile { character -> character == '-' }
      } else {
        option
      }
    }

  for (i in indices) {
    this[i] = normalized[i]
  }
}

private fun MutableList<String>.toggleStartPausedOption(autoplay: Boolean) {
  val hasOption = contains(":start-paused")

  if (!autoplay && !hasOption) {
    add(":start-paused")
  }
}

File: ios/Player/MediaPlayer.swift
import ExpoModulesCore
import UIKit
import VLCKit

private let dialogCustomUI: Bool = true

class MediaPlayer: NSObject {
  private unowned let view: LibVlcPlayerView

  private let video: MediaPlayerDrawable = .init()
  private var picture: PictureInPictureDrawable!

  private var library: VLCLibrary?
  var mediaPlayer: VLCMediaPlayer?
  var vlcDialog: VLCDialogProvider?
  var vlcDialogRef: NSValue?

  var userStop: Bool = false
  private var firstPlay: Bool = true
  var shouldInit: Bool = true

  var hasVideoSize: Bool {
    let video = getVideo()
    return video.width > 0 && video.height > 0
  }

  private var hasMediaLength: Bool {
    let length = getLength()
    return length > 0
  }

  private var hasMediaVolume: Bool {
    let volume = mediaPlayer?.audio?.volume ?? Int32(MediaPlayerConstants.minPlayerVolume)
    return volume > MediaPlayerConstants.minPlayerVolume
  }

  init(_ view: LibVlcPlayerView) {
    self.view = view
    picture = PictureInPictureDrawable(view)
    super.init()
  }

  func applyBounds() {
    video.transform = .identity
    video.frame = view.bounds
    picture.transform = .identity
    picture.frame = view.bounds
  }

  func initPlayer() {
    if shouldInit {
      destroyPlayer()

      if view.source != nil {
        createPlayer()
      }
    }
  }

  private func createPlayer() {
    let drawable = view.pictureInPicture
      ? picture!
      : video

    library = VLCLibrary()
    mediaPlayer = VLCMediaPlayer(library: library!)
    mediaPlayer!.drawable = drawable
    mediaPlayer!.delegate = self
    addPlayerSlaves(view.slaves)

    vlcDialog = VLCDialogProvider(library: library!, customUI: dialogCustomUI)
    vlcDialog!.customRenderer = self

    guard let source = view.source, let url = URL(string: source) else {
      view.onEncounteredError(["message": "Invalid source, media could not be set"])
      return
    }

    var args = view.options
    args.normalizeOptions()
    args.toggleStartPausedOption(view.autoplay)

    let media = VLCMedia(url: url)
    args.forEach { arg in media!.addOption(arg) }
    mediaPlayer!.media = media
    mediaPlayer!.play()

    firstPlay = true
    shouldInit = false

    view.subviews.forEach { subview in subview.removeFromSuperview() }
    view.addSubview(drawable)
  }

  func destroyPlayer() {
    library = nil
    mediaPlayer?.stop()
    mediaPlayer = nil
    vlcDialog?.customRenderer = nil
    vlcDialog = nil
  }

  private func setupPlayer() {
    DispatchQueue.main.async { [weak self] in
      guard let self else { return }

      if let player = mediaPlayer {
        if view.scale != MediaPlayerConstants.defaultPlayerScale {
          player.scaleFactor = Float(view.scale)
        }

        if view.rate != MediaPlayerConstants.defaultPlayerRate {
          player.rate = Float(view.rate)
        }

        if view.time != MediaPlayerConstants.defaultPlayerTime {
          player.time = VLCTime(int: Int32(view.time))
        }

        // Negative volume workaround
        retryUntil { [weak self] _ in
          guard let self else { return true }

          let newVolume = view.mute ?
            MediaPlayerConstants.minPlayerVolume :
            view.volume

          player.audio?.volume = Int32(newVolume)

          return false
        }

        view.time = MediaPlayerConstants.defaultPlayerTime
      }
    }
  }

  func addPlayerSlaves(_ slaves: [Slave]) {
    for slave in slaves {
      let source = slave.source
      let type = slave.type
      let slaveType = type == "subtitle" ?
        VLCMediaPlaybackSlaveType.subtitle :
        VLCMediaPlaybackSlaveType.audio
      let selected = slave.selected ?? false

      guard let url = URL(string: source) else {
        view.onEncounteredError(["message": "Invalid source, \(type) could not be added"])
        continue
      }

      mediaPlayer?.addPlaybackSlave(url, type: slaveType, enforce: selected)
    }
  }

  private func selectTrack(_ index: Int, _ type: VLCMedia.TrackType) {
    if let player = mediaPlayer {
      if index == -1 {
        switch type {
        case .audio: player.deselectAllAudioTracks()
        case .video: player.deselectAllVideoTracks()
        case .text: player.deselectAllTextTracks()
        default: break
        }
      } else {
        player.selectTrack(at: index, type: type)
      }
    }
  }

  func setPlayerTracks() {
    let audioTrack = view.tracks?.audio
    let videoTrack = view.tracks?.video
    let textTrack = view.tracks?.subtitle

    if let audioTrack { selectTrack(audioTrack, .audio) }
    if let videoTrack { selectTrack(videoTrack, .video) }
    if let textTrack { selectTrack(textTrack, .text) }
  }

  func setPlayerDelays() {
    if let player = mediaPlayer {
      let audioDelay = view.delays?.audio
      let textDelay = view.delays?.subtitle

      if let audioDelay { player.currentAudioPlaybackDelay = NSInteger(audioDelay) }
      if let textDelay { player.currentVideoSubTitleDelay = NSInteger(textDelay) }
    }
  }

  private func setContentFit(drawable: MediaPlayerDrawable) {
    DispatchQueue.main.async { [weak self] in
      guard let self else { return }

      var transform: CGAffineTransform = .identity

      let video = getVideo()

      if hasVideoSize == true {
        let viewAspect = drawable.frame.size.width / drawable.frame.size.height
        let videoAspect = CGFloat(video.width) / CGFloat(video.height)

        switch view.contentFit {
        case .contain:
          // No transformation required
          break
        case .cover:
          let scale = videoAspect > viewAspect ?
            videoAspect / viewAspect :
            viewAspect / videoAspect

          transform = CGAffineTransform(scaleX: scale, y: scale)
        case .fill:
          var scaleX = 1.0
          var scaleY = 1.0

          if videoAspect > viewAspect {
            scaleY = videoAspect / viewAspect
          } else {
            scaleX = viewAspect / videoAspect
          }

          transform = CGAffineTransform(scaleX: scaleX, y: scaleY)
        }
      }

      drawable.transform = transform
    }
  }

  func applyContentFit() {
    setContentFit(drawable: video)
    setContentFit(drawable: picture)
  }

  private func getMediaTracks() -> MediaTracks {
    guard let player = mediaPlayer else { return MediaTracks() }

    let disableTrack = MediaTrack(id: -1, name: "Disable")

    let audios = player.audioTracks.enumerated()
    let videos = player.videoTracks.enumerated()
    let subtitles = player.textTracks.enumerated()

    let audio = [disableTrack] + audios.map { index, audio in MediaTrack(
      id: index,
      name: audio.trackName
    ) }
    let video = [disableTrack] + videos.map { index, video in MediaTrack(
      id: index,
      name: video.trackName
    ) }
    let subtitle = [disableTrack] + subtitles.map { index, subtitle in MediaTrack(
      id: index,
      name: subtitle.trackName
    ) }

    return MediaTracks(
      audio: audio,
      video: video,
      subtitle: subtitle
    )
  }

  private func getMedia() -> Media {
    let length = Int(mediaPlayer?.media?.length.intValue ?? 0)
    let seekable = mediaPlayer?.isSeekable ?? false

    return Media(
      length: length,
      seekable: seekable
    )
  }

  private func getMetadata() -> Metadata {
    guard let metaData = mediaPlayer?.media?.metaData else {
      return Metadata()
    }

    let title = metaData.title
    let artist = metaData.artist
    let album = metaData.album
    let artworkURL = metaData.artworkURL?.absoluteString

    return Metadata(
      title: title,
      artist: artist,
      album: album,
      artworkURL: artworkURL
    )
  }

  private func getVideo() -> Video {
    guard let track = mediaPlayer?.videoTracks.first(where: { track in track.isSelected }),
          let video = track.video
    else {
      return Video()
    }

    let width = Int(video.width)
    let height = Int(video.height)
    let frameRate = video.frameRateDenominator != 0 ?
      Int(video.frameRate / video.frameRateDenominator) :
      0
    let bitrate = Int(track.bitrate)

    return Video(
      width: width,
      height: height,
      frameRate: frameRate,
      bitrate: bitrate
    )
  }

  private func getMediaInfo() -> MediaInfo {
    let media = getMedia()
    let metadata = getMetadata()
    let video = getVideo()

    return MediaInfo(
      media: media,
      metadata: metadata,
      video: video
    )
  }

  private func getLength() -> Int {
    Int(mediaPlayer?.media?.length.intValue ?? 0)
  }

  func startPictureInPicture() throws {
    try picture.startPictureInPicture()
  }

  func stopPictureInPicture() {
    picture.stopPictureInPicture()
  }

  private func retryUntil(
    maxRetries: Int = MediaPlayerConstants.maxRetryCount,
    retry: Int = 0,
    delay: Double = MediaPlayerConstants.retryDelayMs,
    block: @escaping (_ isLastAttempt: Bool) -> Bool
  ) {
    let isLastAttempt = retry > maxRetries

    if block(isLastAttempt) || isLastAttempt { return }

    let wait = retry > 0 ? delay : 0
    let nextDelay = retry > 0 ? delay * MediaPlayerConstants.expDelayMultiplier : delay
    let deadline = DispatchTime.now() + DispatchTimeInterval.milliseconds(Int(wait))

    DispatchQueue.main.asyncAfter(deadline: deadline) { [weak self] in
      self?.retryUntil(
        maxRetries: maxRetries,
        retry: retry + 1,
        delay: nextDelay,
        block: block
      )
    }
  }
}

extension MediaPlayer: VLCMediaPlayerDelegate {
  func mediaPlayerStateChanged(_ newState: VLCMediaPlayerState) {
    guard let player = mediaPlayer else { return }

    switch newState {
    case .playing,
         .paused,
         .stopped:
      if newState == .playing {
        view.onPlaying()

        if firstPlay {
          setupPlayer()
          setPlayerTracks()
          setPlayerDelays()

          retryUntil { [weak self] _ in
            guard let self else { return true }

            if hasVideoSize {
              applyContentFit()
            }

            return hasVideoSize
          }

          retryUntil { [weak self] _ in
            guard let self else { return true }

            if hasMediaVolume {
              MediaPlayerManager.shared.audioSessionManager
                .setAppropriateAudioSession()
            }

            return hasMediaVolume
          }

          retryUntil { [weak self] isLastAttempt in
            guard let self else { return true }

            if hasMediaLength || isLastAttempt {
              view.onFirstPlay(getMediaInfo())
            }

            return hasMediaLength
          }

          firstPlay = false
        }
      }

      if newState == .paused {
        view.onPaused()
      }

      if newState == .stopped {
        view.onStopped()

        if view.Repeat, !userStop {
          player.play()
        }

        userStop = false
        firstPlay = true
      }

      MediaPlayerManager.shared.keepAwakeManager.toggleKeepAwake()
      MediaPlayerManager.shared.audioSessionManager.setAppropriateAudioSession()
      picture.updatePipState()
    case .error:
      view.onEncounteredError(["message": "Player encountered an error"])

      player.stop()
    default:
      break
    }
  }

  func mediaPlayerBufferingChanged(_ buffering: Float) {
    view.onBuffering(["value": buffering])
  }

  func mediaPlayerLengthChanged(_: Int64) {
    picture.updatePipState()
  }

  func mediaPlayerTimeChanged(_: Notification) {
    guard let player = mediaPlayer else { return }

    view.onTimeChanged(["value": player.time.intValue])
    view.onPositionChanged(["value": player.position])
  }

  func mediaPlayerTrackAdded(_: String, with _: VLCMedia.TrackType) {
    view.onESAdded(getMediaTracks())
  }

  func mediaPlayerStartedRecording(_: VLCMediaPlayer) {
    let recording = Recording(
      path: nil,
      isRecording: true
    )

    view.onRecordChanged(recording)
  }

  func mediaPlayer(recordingStoppedAt path: String) {
    let recording = Recording(
      path: path,
      isRecording: false
    )

    view.onRecordChanged(recording)
  }
}

extension MediaPlayer: VLCCustomDialogRendererProtocol {
  func showError(
    withTitle title: String,
    message: String
  ) {
    let dialog = Dialog(
      title: title,
      text: message,
      type: "error"
    )

    view.onDialogDisplay(dialog)
  }

  func showLogin(
    withTitle title: String,
    message: String,
    defaultUsername _: String?,
    askingForStorage _: Bool,
    withReference reference: NSValue
  ) {
    vlcDialogRef = reference

    let dialog = Dialog(
      title: title,
      text: message,
      type: "login"
    )

    view.onDialogDisplay(dialog)
  }

  func showQuestion(
    withTitle title: String,
    message: String,
    type _: VLCDialogQuestionType,
    cancel: String?,
    action1String: String?,
    action2String: String?,
    withReference reference: NSValue
  ) {
    vlcDialogRef = reference

    let dialog = Dialog(
      title: title,
      text: message,
      type: "question",
      cancelText: cancel,
      action1Text: action1String,
      action2Text: action2String
    )

    view.onDialogDisplay(dialog)
  }

  func showProgress(
    withTitle _: String,
    message _: String,
    isIndeterminate _: Bool,
    position _: Float,
    cancel _: String?,
    withReference _: NSValue
  ) {}

  func updateProgress(
    withReference _: NSValue,
    message _: String?,
    position _: Float
  ) {}

  func cancelDialog(withReference _: NSValue) {}
}

private extension [String] {
  mutating func normalizeOptions() {
    self = map { option in
      if !option.hasPrefix(":") {
        ":" + option.drop { character in character == "-" }
      } else {
        option
      }
    }
  }
}

private extension [String] {
  mutating func toggleStartPausedOption(_ autoplay: Bool) {
    let hasOption = self.contains(":start-paused")

    if !autoplay, !hasOption {
      append(":start-paused")
    }
  }
}
